Type bounds fails to find Deref is implemented for i32

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

After fixing #2190 i was able to find our Type Bounds checking code is unable to realise the generic reference impl blocks do apply to i32 here. I think Rustc uses lang items here to say yes this impl block is applied to this type but i need to double check how they do this. We could reuse an autoderef cycle to match here as well for i32 vs &T from our coercion code.

I tried this code:

```rust
#[lang = "deref"]
trait Deref {
    type Target;
    fn deref(&self) -> &Self::Target;
}

impl<T> Deref for &T {
    type Target = T;

    fn deref(&self) -> &T {
        *self
    }
}

impl i32 {
    fn max(self, other: i32) -> i32 {
        if self > other {
            self
        } else {
            other
        }
    }
}

fn foo<T: Deref<Target = i32>>(t: &T) -> i32 {
    t.max(2)
}

fn main() -> i32 {
    let a: i32 = 1;
    foo::<i32>(&a) - 2
}
```

I expected to see this happen: Compile

Instead, this happened: 

```
test.rs:32:11: error: bounds not satisfied for i32 ‘Deref’ is not satisfied
   26 | fn foo<T: Deref<Target = i32>>(t: &T) -> i32 {
      |           ~~~~~
......
   32 |     foo::<i32>(&a) - 2
      |           ^~~
```

### Meta

- What version of Rust GCC were you using, git sha if possible. 586d85e4947d36852faac4c27229e1ba3205e270

--- a/gcc/rust/backend/rust-compile-base.cc
+++ b/gcc/rust/backend/rust-compile-base.cc
@@ -766,27 +766,28 @@ HIRCompileBase::named_constant_expression (tree type_tree,
 }
 
 tree
-HIRCompileBase::resolve_method_address (
-  TyTy::FnType *fntype, HirId ref, TyTy::BaseType *receiver,
-  const HIR::PathIdentSegment &segment,
-  const Analysis::NodeMapping &expr_mappings, Location expr_locus)
+HIRCompileBase::resolve_method_address (TyTy::FnType *fntype,
+					TyTy::BaseType *receiver,
+					Location expr_locus)
 {
+  DefId id = fntype->get_id ();
+  rust_assert (id != UNKNOWN_DEFID);
+
   // Now we can try and resolve the address since this might be a forward
   // declared function, generic function which has not be compiled yet or
   // its an not yet trait bound function
-  HIR::ImplItem *resolved_item
-    = ctx->get_mappings ()->lookup_hir_implitem (ref, nullptr);
+  HIR::Item *resolved_item = ctx->get_mappings ()->lookup_defid (id);
   if (resolved_item != nullptr)
     {
       if (!fntype->has_subsititions_defined ())
-	return CompileInherentImplItem::Compile (resolved_item, ctx);
+	return CompileItem::compile (resolved_item, ctx);
 
-      return CompileInherentImplItem::Compile (resolved_item, ctx, fntype);
+      return CompileItem::compile (resolved_item, ctx, fntype);
     }
 
   // it might be resolved to a trait item
   HIR::TraitItem *trait_item
-    = ctx->get_mappings ()->lookup_hir_trait_item (ref);
+    = ctx->get_mappings ()->lookup_trait_item_defid (id);
   HIR::Trait *trait = ctx->get_mappings ()->lookup_trait_item_mapping (
     trait_item->get_mappings ().get_hirid ());
 
@@ -800,7 +801,7 @@ HIRCompileBase::resolve_method_address (
   // item so its up to us to figure out if this path should resolve
   // to an trait-impl-block-item or if it can be defaulted to the
   // trait-impl-item's definition
-
+  const HIR::PathIdentSegment segment (trait_item->trait_identifier ());
   auto root = receiver->get_root ();
   auto candidates
     = Resolver::PathProbeImplTrait::Probe (root, segment, trait_ref);
@@ -840,7 +841,8 @@ HIRCompileBase::resolve_method_address (
       TyTy::BaseType *infer_impl_call
 	= candidate_call->infer_substitions (expr_locus);
       monomorphized
-	= Resolver::unify_site (ref, TyTy::TyWithLocation (infer_impl_call),
+	= Resolver::unify_site (fntype->get_ref (),
+				TyTy::TyWithLocation (infer_impl_call),
 				TyTy::TyWithLocation (fntype), expr_locus);
     }
 

--- a/gcc/rust/backend/rust-compile-base.h
+++ b/gcc/rust/backend/rust-compile-base.h
@@ -80,10 +80,7 @@ protected:
   tree resolve_unsized_dyn_adjustment (Resolver::Adjustment &adjustment,
 				       tree expression, Location locus);
 
-  tree resolve_method_address (TyTy::FnType *fntype, HirId ref,
-			       TyTy::BaseType *receiver,
-			       const HIR::PathIdentSegment &segment,
-			       const Analysis::NodeMapping &expr_mappings,
+  tree resolve_method_address (TyTy::FnType *fntype, TyTy::BaseType *receiver,
 			       Location expr_locus);
 
   void compile_function_body (tree fndecl, HIR::BlockExpr &function_body,

--- a/gcc/rust/backend/rust-compile-expr.cc
+++ b/gcc/rust/backend/rust-compile-expr.cc
@@ -1757,9 +1757,7 @@ CompileExpr::visit (HIR::MethodCallExpr &expr)
       // lookup compiled functions since it may have already been compiled
       HIR::PathExprSegment method_name = expr.get_method_name ();
       HIR::PathIdentSegment segment_name = method_name.get_segment ();
-      fn_expr
-	= resolve_method_address (fntype, ref, receiver, segment_name,
-				  expr.get_mappings (), expr.get_locus ());
+      fn_expr = resolve_method_address (fntype, receiver, expr.get_locus ());
     }
 
   // lookup the autoderef mappings
@@ -1899,9 +1897,7 @@ CompileExpr::resolve_operator_overload (
   // lookup compiled functions since it may have already been compiled
   HIR::PathIdentSegment segment_name (
     Analysis::RustLangItem::ToString (lang_item_type));
-  tree fn_expr
-    = resolve_method_address (fntype, ref, receiver, segment_name,
-			      expr.get_mappings (), expr.get_locus ());
+  tree fn_expr = resolve_method_address (fntype, receiver, expr.get_locus ());
 
   // lookup the autoderef mappings
   std::vector<Resolver::Adjustment> *adjustments = nullptr;

--- a/gcc/rust/backend/rust-compile-expr.cc
+++ b/gcc/rust/backend/rust-compile-expr.cc
@@ -2367,15 +2367,8 @@ HIRCompileBase::resolve_deref_adjustment (Resolver::Adjustment &adjustment,
   rust_assert (adjustment.has_operator_overload ());
 
   TyTy::FnType *lookup = adjustment.get_deref_operator_fn ();
-  HIR::ImplItem *resolved_item = adjustment.get_deref_hir_item ();
-
-  tree fn_address = error_mark_node;
-  if (!lookup->has_subsititions_defined ())
-    fn_address = CompileInherentImplItem::Compile (resolved_item, ctx, nullptr,
-						   true, locus);
-  else
-    fn_address = CompileInherentImplItem::Compile (resolved_item, ctx, lookup,
-						   true, locus);
+  TyTy::BaseType *receiver = adjustment.get_actual ();
+  tree fn_address = resolve_method_address (lookup, receiver, locus);
 
   // does it need a reference to call
   tree adjusted_argument = expression;

--- a/gcc/rust/backend/rust-compile-resolve-path.cc
+++ b/gcc/rust/backend/rust-compile-resolve-path.cc
@@ -213,8 +213,7 @@ HIRCompileBase::query_compile (HirId ref, TyTy::BaseType *lookup,
 	      receiver = fn->get_self_type ();
 	      receiver = receiver->destructure ();
 
-	      return resolve_method_address (fn, ref, receiver, final_segment,
-					     mappings, expr_locus);
+	      return resolve_method_address (fn, receiver, expr_locus);
 	    }
 	}
 

--- a/gcc/rust/typecheck/rust-autoderef.cc
+++ b/gcc/rust/typecheck/rust-autoderef.cc
@@ -28,7 +28,7 @@ namespace Resolver {
 static bool
 resolve_operator_overload_fn (
   Analysis::RustLangItem::ItemType lang_item_type, TyTy::BaseType *ty,
-  TyTy::FnType **resolved_fn, HIR::ImplItem **impl_item,
+  TyTy::FnType **resolved_fn,
   Adjustment::AdjustmentType *requires_ref_adjustment);
 
 TyTy::BaseType *
@@ -44,19 +44,18 @@ Adjustment
 Adjuster::try_deref_type (TyTy::BaseType *ty,
 			  Analysis::RustLangItem::ItemType deref_lang_item)
 {
-  HIR::ImplItem *impl_item = nullptr;
   TyTy::FnType *fn = nullptr;
   Adjustment::AdjustmentType requires_ref_adjustment
     = Adjustment::AdjustmentType::ERROR;
   bool operator_overloaded
-    = resolve_operator_overload_fn (deref_lang_item, ty, &fn, &impl_item,
+    = resolve_operator_overload_fn (deref_lang_item, ty, &fn,
 				    &requires_ref_adjustment);
   if (!operator_overloaded)
     {
       return Adjustment::get_error ();
     }
 
-  auto resolved_base = fn->get_return_type ()->clone ();
+  auto resolved_base = fn->get_return_type ()->destructure ();
   bool is_valid_type = resolved_base->get_kind () == TyTy::TypeKind::REF;
   if (!is_valid_type)
     return Adjustment::get_error ();
@@ -81,7 +80,7 @@ Adjuster::try_deref_type (TyTy::BaseType *ty,
     }
 
   return Adjustment::get_op_overload_deref_adjustment (adjustment_type, ty,
-						       ref_base, fn, impl_item,
+						       ref_base, fn,
 						       requires_ref_adjustment);
 }
 
@@ -94,7 +93,7 @@ Adjuster::try_raw_deref_type (TyTy::BaseType *ty)
 
   const TyTy::ReferenceType *ref_base
     = static_cast<const TyTy::ReferenceType *> (ty);
-  auto infered = ref_base->get_base ()->clone ();
+  auto infered = ref_base->get_base ()->destructure ();
 
   return Adjustment (Adjustment::AdjustmentType::INDIRECTION, ty, infered);
 }
@@ -122,8 +121,8 @@ Adjuster::try_unsize_type (TyTy::BaseType *ty)
 
 static bool
 resolve_operator_overload_fn (
-  Analysis::RustLangItem::ItemType lang_item_type, TyTy::BaseType *ty,
-  TyTy::FnType **resolved_fn, HIR::ImplItem **impl_item,
+  Analysis::RustLangItem::ItemType lang_item_type, TyTy::BaseType *lhs,
+  TyTy::FnType **resolved_fn,
   Adjustment::AdjustmentType *requires_ref_adjustment)
 {
   auto context = TypeCheckContext::get ();
@@ -139,35 +138,69 @@ resolve_operator_overload_fn (
   if (!lang_item_defined)
     return false;
 
-  auto segment = HIR::PathIdentSegment (associated_item_name);
-  auto candidates
-    = MethodResolver::Probe (ty, HIR::PathIdentSegment (associated_item_name),
-			     true);
+  // we might be in a static or const context and unknown is fine
+  TypeCheckContextItem current_context = TypeCheckContextItem::get_error ();
+  if (context->have_function_context ())
+    {
+      current_context = context->peek_context ();
+    }
 
-  bool have_implementation_for_lang_item = !candidates.empty ();
+  // this flags stops recurisve calls to try and deref when none is available
+  // which will cause an infinite loop
+  bool autoderef_flag = true;
+  auto segment = HIR::PathIdentSegment (associated_item_name);
+  auto candidates = MethodResolver::Probe (lhs, segment, autoderef_flag);
+
+  // remove any recursive candidates
+  std::set<MethodCandidate> resolved_candidates;
+  for (auto &c : candidates)
+    {
+      const TyTy::BaseType *candidate_type = c.candidate.ty;
+      rust_assert (candidate_type->get_kind () == TyTy::TypeKind::FNDEF);
+
+      const TyTy::FnType &fn
+	= *static_cast<const TyTy::FnType *> (candidate_type);
+
+      DefId current_fn_defid = current_context.get_defid ();
+      bool recursive_candidated = fn.get_id () == current_fn_defid;
+      if (!recursive_candidated)
+	{
+	  resolved_candidates.insert (c);
+	}
+    }
+
+  bool have_implementation_for_lang_item = resolved_candidates.size () > 0;
   if (!have_implementation_for_lang_item)
     return false;
 
-  // multiple candidates?
-  if (candidates.size () > 1)
+  if (resolved_candidates.size () > 1)
     {
-      // error out? probably not for this case
+      // no need to error out as we are just trying to see if there is a fit
       return false;
     }
 
   // Get the adjusted self
-  auto candidate = *candidates.begin ();
-  Adjuster adj (ty);
+  MethodCandidate candidate = *resolved_candidates.begin ();
+  Adjuster adj (lhs);
   TyTy::BaseType *adjusted_self = adj.adjust_type (candidate.adjustments);
 
-  // is this the case we are recursive
-  // handle the case where we are within the impl block for this
-  // lang_item otherwise we end up with a recursive operator overload
-  // such as the i32 operator overload trait
-  TypeCheckContextItem fn_context = context->peek_context ();
-  if (fn_context.get_type () == TypeCheckContextItem::ItemType::IMPL_ITEM)
+  PathProbeCandidate &resolved_candidate = candidate.candidate;
+  TyTy::BaseType *lookup_tyty = candidate.candidate.ty;
+  rust_assert (lookup_tyty->get_kind () == TyTy::TypeKind::FNDEF);
+  TyTy::BaseType *lookup = lookup_tyty;
+  TyTy::FnType *fn = static_cast<TyTy::FnType *> (lookup);
+  rust_assert (fn->is_method ());
+
+  rust_debug ("is_impl_item_candidate: %s",
+	      resolved_candidate.is_impl_candidate () ? "true" : "false");
+
+  // in the case where we resolve to a trait bound we have to be careful we are
+  // able to do so there is a case where we are currently resolving the deref
+  // operator overload function which is generic and this might resolve to the
+  // trait item of deref which is not valid as its just another recursive case
+  if (current_context.get_type () == TypeCheckContextItem::ItemType::IMPL_ITEM)
     {
-      auto &impl_item = fn_context.get_impl_item ();
+      auto &impl_item = current_context.get_impl_item ();
       HIR::ImplBlock *parent = impl_item.first;
       HIR::Function *fn = impl_item.second;
 
@@ -201,24 +234,17 @@ resolve_operator_overload_fn (
 	}
     }
 
-  TyTy::BaseType *lookup_tyty = candidate.candidate.ty;
-
-  // rust only support impl item deref operator overloading ie you must have an
-  // impl block for it
-  rust_assert (candidate.candidate.type
-	       == PathProbeCandidate::CandidateType::IMPL_FUNC);
-  *impl_item = candidate.candidate.item.impl.impl_item;
-
-  rust_assert (lookup_tyty->get_kind () == TyTy::TypeKind::FNDEF);
-  TyTy::BaseType *lookup = lookup_tyty;
-  TyTy::FnType *fn = static_cast<TyTy::FnType *> (lookup);
-  rust_assert (fn->is_method ());
+  // we found a valid operator overload
+  fn->prepare_higher_ranked_bounds ();
+  rust_debug ("resolved operator overload to: {%u} {%s}",
+	      candidate.candidate.ty->get_ref (),
+	      candidate.candidate.ty->debug_str ().c_str ());
 
   if (fn->needs_substitution ())
     {
-      if (ty->get_kind () == TyTy::TypeKind::ADT)
+      if (lhs->get_kind () == TyTy::TypeKind::ADT)
 	{
-	  const TyTy::ADTType *adt = static_cast<const TyTy::ADTType *> (ty);
+	  const TyTy::ADTType *adt = static_cast<const TyTy::ADTType *> (lhs);
 
 	  auto s = fn->get_self_type ()->get_root ();
 	  rust_assert (s->can_eq (adt, false));
@@ -265,8 +291,8 @@ resolve_operator_overload_fn (
 	  rust_assert (lookup->get_kind () == TyTy::TypeKind::FNDEF);
 	  fn = static_cast<TyTy::FnType *> (lookup);
 
-	  Location unify_locus = mappings->lookup_location (ty->get_ref ());
-	  unify_site (ty->get_ref (),
+	  Location unify_locus = mappings->lookup_location (lhs->get_ref ());
+	  unify_site (lhs->get_ref (),
 		      TyTy::TyWithLocation (fn->get_self_type ()),
 		      TyTy::TyWithLocation (adjusted_self), unify_locus);
 

--- a/gcc/rust/typecheck/rust-autoderef.h
+++ b/gcc/rust/typecheck/rust-autoderef.h
@@ -42,18 +42,15 @@ public:
   // ctor for all adjustments except derefs
   Adjustment (AdjustmentType type, TyTy::BaseType *actual,
 	      TyTy::BaseType *expected)
-    : Adjustment (type, actual, expected, nullptr, nullptr,
-		  AdjustmentType::ERROR)
+    : Adjustment (type, actual, expected, nullptr, AdjustmentType::ERROR)
   {}
 
   static Adjustment get_op_overload_deref_adjustment (
     AdjustmentType type, TyTy::BaseType *actual, TyTy::BaseType *expected,
-    TyTy::FnType *fn, HIR::ImplItem *deref_item,
-    Adjustment::AdjustmentType requires_ref_adjustment)
+    TyTy::FnType *fn, Adjustment::AdjustmentType requires_ref_adjustment)
   {
     rust_assert (type == DEREF || type == DEREF_MUT);
-    return Adjustment (type, actual, expected, fn, deref_item,
-		       requires_ref_adjustment);
+    return Adjustment (type, actual, expected, fn, requires_ref_adjustment);
   }
 
   AdjustmentType get_type () const { return type; }
@@ -107,15 +104,12 @@ public:
     return requires_ref_adjustment;
   }
 
-  HIR::ImplItem *get_deref_hir_item () const { return deref_item; }
-
 private:
   Adjustment (AdjustmentType type, TyTy::BaseType *actual,
 	      TyTy::BaseType *expected, TyTy::FnType *deref_operator_fn,
-	      HIR::ImplItem *deref_item,
 	      Adjustment::AdjustmentType requires_ref_adjustment)
     : type (type), actual (actual), expected (expected),
-      deref_operator_fn (deref_operator_fn), deref_item (deref_item),
+      deref_operator_fn (deref_operator_fn),
       requires_ref_adjustment (requires_ref_adjustment)
   {}
 
@@ -127,7 +121,6 @@ private:
   //
   // the fn that we are calling
   TyTy::FnType *deref_operator_fn;
-  HIR::ImplItem *deref_item;
   // operator overloads can requre a reference
   Adjustment::AdjustmentType requires_ref_adjustment;
 };

--- a/gcc/rust/typecheck/rust-hir-dot-operator.cc
+++ b/gcc/rust/typecheck/rust-hir-dot-operator.cc
@@ -66,8 +66,9 @@ MethodResolver::try_hook (const TyTy::BaseType &r)
 bool
 MethodResolver::select (TyTy::BaseType &receiver)
 {
-  rust_debug ("MethodResolver::select reciever=[%s]",
-	      receiver.debug_str ().c_str ());
+  rust_debug ("MethodResolver::select reciever=[%s] path=[%s]",
+	      receiver.debug_str ().c_str (),
+	      segment_name.as_string ().c_str ());
 
   struct impl_item_candidate
   {

--- a/gcc/rust/typecheck/rust-hir-type-check-path.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-path.cc
@@ -422,6 +422,11 @@ TypeCheckExpr::resolve_segments (NodeId root_resolved_node_id,
 				     TyTy::TyWithLocation (prev_segment),
 				     TyTy::TyWithLocation (impl_block_ty),
 				     seg.get_locus ());
+	  bool ok = prev_segment->get_kind () != TyTy::TypeKind::ERROR;
+	  if (!ok)
+	    {
+	      return;
+	    }
 
 	  if (found_impl_trait)
 	    {

--- a/gcc/rust/typecheck/rust-tyty-bounds.cc
+++ b/gcc/rust/typecheck/rust-tyty-bounds.cc
@@ -111,7 +111,6 @@ TypeBoundsProbe::assemble_sized_builtin ()
     case TyTy::REF:
     case TyTy::POINTER:
     case TyTy::PARAM:
-    case TyTy::ARRAY:
     case TyTy::SLICE:
     case TyTy::FNDEF:
     case TyTy::FNPTR:
@@ -123,16 +122,16 @@ TypeBoundsProbe::assemble_sized_builtin ()
     case TyTy::FLOAT:
     case TyTy::USIZE:
     case TyTy::ISIZE:
+    case TyTy::CLOSURE:
+    case TyTy::INFER:
       assemble_builtin_candidate (Analysis::RustLangItem::SIZED);
       break;
 
-      // not-sure about this.... FIXME
-    case TyTy::INFER:
+    case TyTy::ARRAY:
     case TyTy::NEVER:
     case TyTy::PLACEHOLDER:
     case TyTy::PROJECTION:
     case TyTy::DYNAMIC:
-    case TyTy::CLOSURE:
     case TyTy::ERROR:
       break;
     }

--- a/gcc/rust/typecheck/rust-tyty.cc
+++ b/gcc/rust/typecheck/rust-tyty.cc
@@ -273,8 +273,11 @@ BaseType::get_locus () const
 
 // FIXME this is missing locus
 bool
-BaseType::satisfies_bound (const TypeBoundPredicate &predicate) const
+BaseType::satisfies_bound (const TypeBoundPredicate &predicate,
+			   bool emit_error) const
 {
+  bool is_infer_var = destructure ()->get_kind () == TyTy::TypeKind::INFER;
+
   const Resolver::TraitReference *query = predicate.get ();
   for (const auto &bound : specified_bounds)
     {
@@ -282,6 +285,9 @@ BaseType::satisfies_bound (const TypeBoundPredicate &predicate) const
       if (item->satisfies_bound (*query))
 	return true;
     }
+
+  if (is_infer_var)
+    return true;
 
   bool satisfied = false;
   auto probed = Resolver::TypeBoundsProbe::Probe (this);
@@ -313,6 +319,11 @@ BaseType::satisfies_bound (const TypeBoundPredicate &predicate) const
       const HIR::ImplBlock &impl = *(b.second);
       for (const auto &item : impl.get_impl_items ())
 	{
+	  bool is_associated_type = item->get_impl_item_type ()
+				    == HIR::ImplItem::ImplItemType::TYPE_ALIAS;
+	  if (!is_associated_type)
+	    continue;
+
 	  TyTy::BaseType *impl_item_ty = nullptr;
 	  Analysis::NodeMapping i = item->get_impl_mappings ();
 	  bool query_ok = Resolver::query_type (i.get_hirid (), &impl_item_ty);
@@ -331,15 +342,21 @@ BaseType::satisfies_bound (const TypeBoundPredicate &predicate) const
 	  // compare the types
 	  if (!bound_ty->can_eq (impl_item_ty, false))
 	    {
-	      RichLocation r (mappings->lookup_location (get_ref ()));
-	      r.add_range (predicate.get_locus ());
-	      r.add_range (mappings->lookup_location (i.get_hirid ()));
+	      if (!impl_item_ty->can_eq (bound_ty, false))
+		{
+		  if (emit_error)
+		    {
+		      RichLocation r (mappings->lookup_location (get_ref ()));
+		      r.add_range (predicate.get_locus ());
+		      r.add_range (mappings->lookup_location (i.get_hirid ()));
 
-	      rust_error_at (
-		r, "expected %<%s%> got %<%s%>",
-		bound_ty->destructure ()->get_name ().c_str (),
-		impl_item_ty->destructure ()->get_name ().c_str ());
-	      return false;
+		      rust_error_at (
+			r, "expected %<%s%> got %<%s%>",
+			bound_ty->destructure ()->get_name ().c_str (),
+			impl_item_ty->destructure ()->get_name ().c_str ());
+		    }
+		  return false;
+		}
 	    }
 	}
 
@@ -357,7 +374,7 @@ BaseType::bounds_compatible (const BaseType &other, Location locus,
     unsatisfied_bounds;
   for (auto &bound : get_specified_bounds ())
     {
-      if (!other.satisfies_bound (bound))
+      if (!other.satisfies_bound (bound, emit_error))
 	unsatisfied_bounds.push_back (bound);
     }
 

--- a/gcc/rust/typecheck/rust-tyty.h
+++ b/gcc/rust/typecheck/rust-tyty.h
@@ -114,7 +114,8 @@ public:
   //     2. (For functions) have the same signature
   virtual bool is_equal (const BaseType &other) const;
 
-  bool satisfies_bound (const TypeBoundPredicate &predicate) const;
+  bool satisfies_bound (const TypeBoundPredicate &predicate,
+			bool emit_error) const;
   bool bounds_compatible (const BaseType &other, Location locus,
 			  bool emit_error) const;
   void inherit_bounds (const BaseType &other);

--- a/gcc/rust/typecheck/rust-unify.cc
+++ b/gcc/rust/typecheck/rust-unify.cc
@@ -151,13 +151,26 @@ UnifyRules::go ()
 	      rtype->debug_str ().c_str ());
 
   // check bounds
-  if (ltype->num_specified_bounds () > 0)
+  bool should_check_bounds = !ltype->is_equal (*rtype);
+  if (should_check_bounds)
     {
-      if (!ltype->bounds_compatible (*rtype, locus, emit_error))
+      if (ltype->num_specified_bounds () > 0)
 	{
-	  // already emitted an error
-	  emit_error = false;
-	  return new TyTy::ErrorType (0);
+	  if (!ltype->bounds_compatible (*rtype, locus, emit_error))
+	    {
+	      // already emitted an error
+	      emit_error = false;
+	      return new TyTy::ErrorType (0);
+	    }
+	}
+      else if (rtype->num_specified_bounds () > 0)
+	{
+	  if (!rtype->bounds_compatible (*ltype, locus, emit_error))
+	    {
+	      // already emitted an error
+	      emit_error = false;
+	      return new TyTy::ErrorType (0);
+	    }
 	}
     }
 

--- a/gcc/testsuite/rust/compile/issue-2190-1.rs
+++ b/gcc/testsuite/rust/compile/issue-2190-1.rs
@@ -1,0 +1,11 @@
+#[lang = "deref"]
+trait Deref {
+    type Target;
+    fn deref(&self) -> &Self::Target;
+}
+
+fn foo<T: Deref<Target = i32>>(t: &T) -> i32 {
+    t.max(2)
+    // { dg-error "failed to resolve method for .max." "" { target *-*-* } .-1 }
+    // { dg-error "failed to type resolve expression" "" { target *-*-* } .-2 }
+}

--- a/gcc/testsuite/rust/compile/issue-2190-2.rs
+++ b/gcc/testsuite/rust/compile/issue-2190-2.rs
@@ -1,0 +1,20 @@
+// { dg-options "-w" }
+#[lang = "deref"]
+trait Deref {
+    type Target;
+    fn deref(&self) -> &Self::Target;
+}
+
+fn foo<T: Deref<Target = i32>>(t: &T) -> i32 {
+    t.max(2)
+}
+
+impl i32 {
+    fn max(self, other: i32) -> i32 {
+        if self > other {
+            self
+        } else {
+            other
+        }
+    }
+}

--- a/gcc/testsuite/rust/execute/torture/issue-2236.rs
+++ b/gcc/testsuite/rust/execute/torture/issue-2236.rs
@@ -1,0 +1,37 @@
+// { dg-options "-w" }
+mod core {
+    mod ops {
+        #[lang = "deref"]
+        trait Deref {
+            type Target;
+            fn deref(&self) -> &Self::Target;
+        }
+
+        impl<T> Deref for &T {
+            type Target = T;
+
+            fn deref(&self) -> &T {
+                *self
+            }
+        }
+    }
+}
+
+impl i32 {
+    fn max(self, other: i32) -> i32 {
+        if self > other {
+            self
+        } else {
+            other
+        }
+    }
+}
+
+fn foo<T: core::ops::Deref<Target = i32>>(t: T) -> i32 {
+    t.max(2)
+}
+
+fn main() -> i32 {
+    let a: i32 = 1;
+    foo(&a) - 2
+}

